### ISSUE TITLE
History of files should optionally show include logs from a rename (git log --follow)

### ISSUE THREAD

**Issue by sluedecke** [gogits/gogs#2487]
Currently the  history of a renamed / moved file starts after the rename.  It would be great to have an option to show the complete history.

**Comment by Unknwon** [comment]
Thanks your feedback!

Never used this, would you mind given a detailed example by using the `--follow`?

**Comment by sluedecke** [comment]
Sure, just create a git repository as follows and check the history of the file bar.txt both from the command line and after you imported the repository into gogs.  Gogs only shows the history of bar.txt after the renaming

```
#
# Create simple git repository with a single file
#

mkdir git-experiment
cd git-experiment
git init .
echo Hello > foo.txt
git add foo.txt 
git commit -m 'initial commit'
git log foo.txt


#
# now rename foo
#
git mv foo.txt  bar.txt
git commit -m 'renamed foo to bar'


#
# Check history of bar.txt, it's past as foo.txt is revealed only when using --follow
#

git log bar.txt
git log --follow bar.txt

```

**Comment by Unknwon** [comment]
Thanks!

So if not use `--follow`, the history of file `bar.txt` will only show the commit `'renamed foo to bar'`?

**Comment by sluedecke** [comment]
Exactly.

**Comment by fnkr** [comment]
This would have huge performance implications. Git does not store rename operations. Instead, it uses some advanced file comparison algorithms every time you use commands like `git status`, `git show` or `git log --follow`. These are slow.

```
$ time git log README > /dev/null
0,75s user 0,04s system 100% cpu 0,784 total
^^^^^

$ time git log --follow README > /dev/null
8,59s user 0,15s system 100% cpu 8,740 total
^^^^^
```

**Comment by sluedecke** [comment]
Good point. Still if it would be an option it is useful. Like a check box which reloads the page. 

On February 6, 2016 4:31:19 PM GMT+01:00, Florian Kaiser <notifications@github.com> wrote:
>This would have huge performance implications. Git does not store
>rename operations. Instead, it uses some advanced file comparison
>algorithms every time you use commands like `git status`, `git show` or
>`git log --follow`. These are slow.
>
>```
>$ time git log README > /dev/null
>0,75s user 0,04s system 100% cpu 0,784 total
>^^^^^
>
>$ time git log --follow README > /dev/null
>8,59s user 0,15s system 100% cpu 8,740 total
>^^^^^
>```
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/gogits/gogs/issues/2487#issuecomment-180790298

-- 
Sascha Lüdecke
+49 179 9210765

**Comment by jimshowalter** [comment]
Not being able to track history in github web UI makes github web UI basically useless in projects engaging in lots of refactoring. We recently moved about 90% of the files in our 4M-line codebase, and now we can't use your product to investigate bugs, because we can't find the file owners, because you don't display the history. Give us a radio button to follow history, with the understanding that it will be slow. Currently it's infinitely slow, so merely slow would be a huge improvement.

**Comment by kuro68k** [comment]
This is indeed a severe issue where I work, and for personal projects. 

It works fine on BitBucket and GitLab. Currently the only fix is to migrate to those platforms. Fortunately they both provide tools to make the transition easier, and the history of renamed files becomes immediately visible afterwards.
